### PR TITLE
Add daily jvmImpl download counts to stats updater

### PR DESCRIPTION
**Pull request by adam-thorpe**
##### Checklist

- [x] `mvn clean package` build and test completes
- [x] documentation is changed or added
- [x] added a new test for this

##### Features
- Add jvmImplDownloads to GithubStats
- Add jvmImpl and feature_version to DockerStats
- Add StatsCalculatorTest
- Fix bug where Binary.download_count wasn't using Installer download_count

This is part one of https://github.com/AdoptOpenJDK/openjdk-api-v3/issues/129. Once this is done, I can then start working on the front end. Would be nice to have data starting by May.

Signed-off-by: Adam Thorpe <adam.thorpe@ibm.com>

### CODE REVIEW

**Comment by gdams** [comment]
@adam-thorpe more linter failures,

worth downloading `ktlint` on your machine and then running
`ktlint --install-git-pre-commit-hook`

that will auto check your commits to this repo.

Also `ktlint -F` is a useful command to auto-fix issues

**Comment by adam-thorpe** [comment]
Test passes fine locally, not sure why it's refusing to connect to the db. Do I need to inherit the init methods from BaseTest ?

**Comment by johnoliver** [comment]
so you have updated the format of the DB classes:

- Does this need to go hand in hand with a schema upgrade of old db entries?
- Will this work with a mix of new and old DB classes? 
- Will old entries without the new data you have added still show sane data?

**Comment by adam-thorpe** [comment]
To be fair, I'm not to sure what will happen. 
- Since all I'm doing is adding data, not modifying anything, the schema should stay intact.
- I imagine older entries will have no data for these new fields. For Github data, theres no way to go back and add the data to older entries as that data is now lost. However, docker data is only recording jvmImpl which can be easily extracted from the name. So I could potentially go back through the old data and update it..? Not sure how I would do that.

Either way, in the front end I will need to take this into consideration when designing routes, as the data will be limited. I don't see this as too much of a problem as I can put a hard-coded date as the earliest entry if need be.

I need to do a full cycle of the updater tool with these changes to see exactly what happens.

**Comment by adam-thorpe** [comment]
I think this is ready for review @johnoliver. Full update works as intended.

- each entry in DockerStats now has this extra data:
```
"feature_version" : 13,
"jvm_impl" : "openj9",
```
for official, these are null.

- each entry in GithubStats now has this extra data:
```
"jvmImplDownloads" : {
	"hotspot" : NumberLong(65787891),
	"openj9" : NumberLong(599343)
},
```
which sum up to the total download count.

**Comment by adam-thorpe** [comment]
Actually wait. When running /tracking on the frontend, I get "Internal error". Need to investigate but may be because it doesn't know what to do with the new data

File: adoptopenjdk-api-v3-updater/src/main/kotlin/net/adoptopenjdk/api/v3/stats/DockerStatsInterface.kt [hunk 1]
```diff
@@ -40,15 +41,21 @@ class DockerStatsInterface {
 
         return pullAllStats()
                 .map {
-                    DockerDownloadStatsDbEntry(now, it.getJsonNumber("pull_count").longValue(), it.getString("name"))
+                    DockerDownloadStatsDbEntry(
+                        now,
+                        it.getJsonNumber("pull_count").longValue(),
+                        it.getString("name"),
+                        "openjdk[0-9]+".toRegex().find(it.getString("name"))!!.value.substring(7).toInt(),
```
Comment (johnoliver) [inline adoptopenjdk-api-v3-updater/src/main/kotlin/net/adoptopenjdk/api/v3/stats/DockerStatsInterface.kt:48]: fairly complex line, could probably extract it into a method, that 7 also seems kind of brittle, can use a named capture group:
```
val x = "openjdk(?<featureNum>[0-9]+)".toRegex().matchEntire("openjdk99")?.groups?.get("featureNum")?.value?.toInt()
```

File: adoptopenjdk-api-v3-updater/src/main/kotlin/net/adoptopenjdk/api/v3/stats/GithubDownloadStatsCalculator.kt [hunk 2]
```diff
@@ -53,23 +54,40 @@ class GithubDownloadStatsCalculator {
...
+                        jvmImpl to
+                            featureRelease
+                            .releases
+                            .getReleases()
+                            .filter { it.vendor == Vendor.adoptopenjdk }
+                            .sumBy {
+                                it.binaries
+                                .filter { it.jvm_impl == jvmImpl }
+                                .sumBy {
+                                    it.download_count.toInt()
```
Comment (johnoliver) [inline adoptopenjdk-api-v3-updater/src/main/kotlin/net/adoptopenjdk/api/v3/stats/GithubDownloadStatsCalculator.kt:79]: I doubt the toInt and the toLong below are required here

Comment (adam-thorpe) [reply adoptopenjdk-api-v3-updater/src/main/kotlin/net/adoptopenjdk/api/v3/stats/GithubDownloadStatsCalculator.kt:79]: Its just how the typing works, sumBy expects an Int to be returned and I made jvm download counts a long to match the overall download count

**Pull request merged**